src/stitching/addTypenameForFragments.ts

```diff
@@ -0,0 +1,52 @@
+import { ASTNode, GraphQLSchema, visit, FieldNode, Kind } from 'graphql';
+
+export default function addTypenameForFragments<T extends ASTNode>(
+  node: T,
+  schema: GraphQLSchema,
+): T {
+  return visit(node, {
+    [Kind.FIELD]: {
+      leave(field: FieldNode) {
+        if (field.selectionSet) {
+          const selections = field.selectionSet.selections;
+
+          let hasFragment = false;
```

Comment:
We would need to check all 3 - is interface or union, has fragment, and has typename. But that requires visit with typename which is a bit annoying